--- C4/SIP/SIPServer.pm.orig	2020-10-27 17:18:36.000000000 -0400
+++ C4/SIP/SIPServer.pm	2020-11-14 15:50:23.510805561 -0500
@@ -125,6 +125,9 @@
         die "process_request: Bad server connection";
     }
 
+    my $ipaddr = $self->{server}->{client}->peerhost;
+    siplog( 'LOG_DEBUG', "process_request: client IP address = $ipaddr" );
+
     $transport = $transports{$self->{service}->{transport}};
 
     if (!defined($transport)) {
--- C4/SIP/Sip/MsgType.pm.orig	2020-10-27 17:18:36.000000000 -0400
+++ C4/SIP/Sip/MsgType.pm	2020-11-14 15:38:09.179667426 -0500
@@ -19,6 +19,7 @@
 use C4::Auth qw(&check_api_auth);
 
 use Koha::Patron::Attributes;
+use Koha::Plugins::Handler;
 use Koha::Items;
 
 use UNIVERSAL::can;
@@ -488,6 +489,25 @@
     return $resp;
 }
 
+# Return true if the patron is valid for the particular client
+# that issued the SIP2 request.
+
+sub check_patron {
+    my ( $patron, $server ) = @_;
+
+    if ( $patron ) {
+        my $plugin_class = $server->{account}->{plugin_class};
+	if ($plugin_class) {
+	    return Koha::Plugins::Handler->run({
+		    class  => $plugin_class,
+		    method => 'sip2_validate_patron',
+		    params => { patron => $patron, server => $server },
+		});
+	}
+    }
+    return 1;
+}
+
 sub handle_patron_status {
     my ( $self, $server ) = @_;
     my $ils = $server->{ils};
@@ -499,6 +519,7 @@
 
     $ils->check_inst_id( $fields->{ (FID_INST_ID) }, "handle_patron_status" );
     $patron = $ils->find_patron( $fields->{ (FID_PATRON_ID) } );
+    $patron = undef unless ( check_patron ( $patron, $server ) );
     $resp = build_patron_status( $patron, $lang, $fields, $server );
     $self->write_msg( $resp, undef, $server->{account}->{terminator}, $server->{account}->{encoding} );
     return (PATRON_STATUS_REQ);
@@ -960,6 +981,7 @@
     $end          = $fields->{ (FID_END_ITEM) };
 
     $patron = $ils->find_patron($patron_id);
+    $patron = undef unless ( check_patron ( $patron, $server ) );
 
     $resp = (PATRON_INFO_RESP);
     if ($patron) {
